### ISSUE TITLE
Sporadic NCCL error "ring 0 does not loop back to start "

### ISSUE THREAD

**Issue by teng-li** [NVIDIA/nccl#179]
Any ideas on what is causing this?

2 DGX-1 node with 8 V100s per node, 4 GPUs requested from each node. Hitting this error during NCCL broadcast with 4 processes / node

```
2019-02-11 09:47:49,051 46393 INFO     Wrapping distributed manually
learnfair0536:46393:46393 [0] NCCL INFO NET : Using interface enp1s0f0:100.97.70.134<0>
learnfair0536:46393:46393 [0] NCCL INFO NET/IB : Using interface enp1s0f0 for sideband communication
learnfair0536:46393:46393 [0] NCCL INFO NET/IB: [0] mlx5_1:1/IB
learnfair0536:46393:46393 [0] NCCL INFO NET/IB: [1] mlx5_3:1/IB
learnfair0536:46393:46393 [0] NCCL INFO NET/IB: [2] mlx5_0:1/IB
learnfair0536:46393:46393 [0] NCCL INFO NET/IB: [3] mlx5_2:1/IB
learnfair0536:46393:46393 [0] NCCL INFO Using internal Network IB
learnfair0536:46393:46393 [0] NCCL INFO NET : Using interface enp1s0f0:100.97.70.134<0>
learnfair0536:46393:46393 [0] NCCL INFO NET/Socket : 1 interfaces found
NCCL version 2.3.7+cuda9.0
learnfair0536:46393:46393 [0] NCCL INFO rank 0 nranks 8
learnfair0536:46395:46395 [2] NCCL INFO NET : Using interface enp1s0f0:100.97.70.134<0>
learnfair0536:46395:46395 [2] NCCL INFO NET/IB : Using interface enp1s0f0 for sideband communication
learnfair0536:46394:46394 [1] NCCL INFO NET : Using interface enp1s0f0:100.97.70.134<0>
learnfair0536:46394:46394 [1] NCCL INFO NET/IB : Using interface enp1s0f0 for sideband communication
learnfair0536:46396:46396 [3] NCCL INFO NET : Using interface enp1s0f0:100.97.70.134<0>
learnfair0536:46396:46396 [3] NCCL INFO NET/IB : Using interface enp1s0f0 for sideband communication
learnfair0536:46395:46395 [2] NCCL INFO NET/IB: [0] mlx5_1:1/IB
learnfair0536:46394:46394 [1] NCCL INFO NET/IB: [0] mlx5_1:1/IB
learnfair0536:46396:46396 [3] NCCL INFO NET/IB: [0] mlx5_1:1/IB
learnfair0537:27655:27655 [0] NCCL INFO NET : Using interface enp1s0f0:100.97.70.135<0>
learnfair0537:27655:27655 [0] NCCL INFO NET/IB : Using interface enp1s0f0 for sideband communication
learnfair0537:27656:27656 [1] NCCL INFO NET : Using interface enp1s0f0:100.97.70.135<0>
learnfair0537:27656:27656 [1] NCCL INFO NET/IB : Using interface enp1s0f0 for sideband communication
learnfair0537:27657:27657 [2] NCCL INFO NET : Using interface enp1s0f0:100.97.70.135<0>
learnfair0537:27657:27657 [2] NCCL INFO NET/IB : Using interface enp1s0f0 for sideband communication
learnfair0537:27658:27658 [3] NCCL INFO NET : Using interface enp1s0f0:100.97.70.135<0>
learnfair0537:27658:27658 [3] NCCL INFO NET/IB : Using interface enp1s0f0 for sideband communication
learnfair0536:46395:46395 [2] NCCL INFO NET/IB: [1] mlx5_3:1/IB
learnfair0536:46396:46396 [3] NCCL INFO NET/IB: [1] mlx5_3:1/IB
learnfair0536:46394:46394 [1] NCCL INFO NET/IB: [1] mlx5_3:1/IB
learnfair0536:46395:46395 [2] NCCL INFO NET/IB: [2] mlx5_0:1/IB
learnfair0536:46394:46394 [1] NCCL INFO NET/IB: [2] mlx5_0:1/IB
learnfair0536:46396:46396 [3] NCCL INFO NET/IB: [2] mlx5_0:1/IB
learnfair0536:46395:46395 [2] NCCL INFO NET/IB: [3] mlx5_2:1/IB
learnfair0536:46395:46395 [2] NCCL INFO Using internal Network IB
learnfair0536:46395:46395 [2] NCCL INFO rank 4 nranks 8
learnfair0536:46394:46394 [1] NCCL INFO NET/IB: [3] mlx5_2:1/IB
learnfair0536:46396:46396 [3] NCCL INFO NET/IB: [3] mlx5_2:1/IB
learnfair0536:46396:46396 [3] NCCL INFO Using internal Network IB
learnfair0536:46396:46396 [3] NCCL INFO rank 6 nranks 8
learnfair0536:46394:46394 [1] NCCL INFO Using internal Network IB
learnfair0536:46394:46394 [1] NCCL INFO rank 2 nranks 8
learnfair0537:27655:27655 [0] NCCL INFO NET/IB: [0] mlx5_1:1/IB
learnfair0537:27656:27656 [1] NCCL INFO NET/IB: [0] mlx5_1:1/IB
learnfair0537:27657:27657 [2] NCCL INFO NET/IB: [0] mlx5_1:1/IB
learnfair0537:27658:27658 [3] NCCL INFO NET/IB: [0] mlx5_1:1/IB
learnfair0537:27655:27655 [0] NCCL INFO NET/IB: [1] mlx5_3:1/IB
learnfair0537:27657:27657 [2] NCCL INFO NET/IB: [1] mlx5_3:1/IB
learnfair0537:27658:27658 [3] NCCL INFO NET/IB: [1] mlx5_3:1/IB
learnfair0537:27656:27656 [1] NCCL INFO NET/IB: [1] mlx5_3:1/IB
learnfair0537:27655:27655 [0] NCCL INFO NET/IB: [2] mlx5_0:1/IB
learnfair0537:27656:27656 [1] NCCL INFO NET/IB: [2] mlx5_0:1/IB
learnfair0537:27658:27658 [3] NCCL INFO NET/IB: [2] mlx5_0:1/IB
learnfair0537:27657:27657 [2] NCCL INFO NET/IB: [2] mlx5_0:1/IB
learnfair0537:27655:27655 [0] NCCL INFO NET/IB: [3] mlx5_2:1/IB
learnfair0537:27658:27658 [3] NCCL INFO NET/IB: [3] mlx5_2:1/IB
learnfair0537:27658:27658 [3] NCCL INFO Using internal Network IB
learnfair0537:27658:27658 [3] NCCL INFO rank 7 nranks 8
learnfair0537:27655:27655 [0] NCCL INFO Using internal Network IB
learnfair0537:27655:27655 [0] NCCL INFO rank 1 nranks 8
learnfair0537:27656:27656 [1] NCCL INFO NET/IB: [3] mlx5_2:1/IB
learnfair0537:27656:27656 [1] NCCL INFO Using internal Network IB
learnfair0537:27656:27656 [1] NCCL INFO rank 3 nranks 8
learnfair0537:27657:27657 [2] NCCL INFO NET/IB: [3] mlx5_2:1/IB
learnfair0537:27657:27657 [2] NCCL INFO Using internal Network IB
learnfair0537:27657:27657 [2] NCCL INFO rank 5 nranks 8
learnfair0536:46393:46747 [0] NCCL INFO comm 0x7fd2500442c0 rank 0 nranks 8
learnfair0536:46396:46761 [3] NCCL INFO comm 0x7f9e6c0442c0 rank 6 nranks 8
learnfair0536:46394:46762 [1] NCCL INFO comm 0x7fc4040442c0 rank 2 nranks 8
learnfair0536:46396:46761 [3] NCCL INFO NET : Using interface enp1s0f0:100.97.70.134<0>
learnfair0536:46396:46761 [3] NCCL INFO NET/Socket : 1 interfaces found
learnfair0536:46394:46762 [1] NCCL INFO NET : Using interface enp1s0f0:100.97.70.134<0>
learnfair0536:46394:46762 [1] NCCL INFO NET/Socket : 1 interfaces found
learnfair0537:27657:27989 [2] NCCL INFO comm 0x7fe11c037b60 rank 5 nranks 8
learnfair0537:27657:27989 [2] NCCL INFO NET : Using interface enp1s0f0:100.97.70.135<0>
learnfair0537:27657:27989 [2] NCCL INFO NET/Socket : 1 interfaces found
learnfair0536:46395:46758 [2] NCCL INFO comm 0x7fdcac0442c0 rank 4 nranks 8
learnfair0536:46395:46758 [2] NCCL INFO NET : Using interface enp1s0f0:100.97.70.134<0>
learnfair0536:46395:46758 [2] NCCL INFO NET/Socket : 1 interfaces found
learnfair0537:27658:27984 [3] NCCL INFO comm 0x7fd734037b60 rank 7 nranks 8
learnfair0537:27655:27985 [0] NCCL INFO comm 0x7f386c037b60 rank 1 nranks 8
learnfair0537:27655:27985 [0] NCCL INFO NET : Using interface enp1s0f0:100.97.70.135<0>
learnfair0537:27655:27985 [0] NCCL INFO NET/Socket : 1 interfaces found
learnfair0537:27658:27984 [3] NCCL INFO NET : Using interface enp1s0f0:100.97.70.135<0>
learnfair0537:27658:27984 [3] NCCL INFO NET/Socket : 1 interfaces found
learnfair0537:27656:27988 [1] NCCL INFO comm 0x7f2194037b60 rank 3 nranks 8
learnfair0537:27656:27988 [1] NCCL INFO NET : Using interface enp1s0f0:100.97.70.135<0>
learnfair0537:27656:27988 [1] NCCL INFO NET/Socket : 1 interfaces found
learnfair0537:27655:27985 [0] NCCL INFO CUDA Dev 0, IB Ports : mlx5_1/1(PHB) mlx5_3/1(SOC) mlx5_0/1(PIX) mlx5_2/1(SOC)
learnfair0537:27656:27988 [1] NCCL INFO CUDA Dev 1, IB Ports : mlx5_1/1(PHB) mlx5_3/1(SOC) mlx5_0/1(PIX) mlx5_2/1(SOC)
learnfair0536:46394:46762 [1] NCCL INFO CUDA Dev 1, IB Ports : mlx5_1/1(SOC) mlx5_3/1(PHB) mlx5_0/1(SOC) mlx5_2/1(PIX)
learnfair0536:46395:46758 [2] NCCL INFO CUDA Dev 2, IB Ports : mlx5_1/1(SOC) mlx5_3/1(PIX) mlx5_0/1(SOC) mlx5_2/1(PHB)
learnfair0537:27657:27989 [2] NCCL INFO CUDA Dev 2, IB Ports : mlx5_1/1(PIX) mlx5_3/1(SOC) mlx5_0/1(PHB) mlx5_2/1(SOC)
learnfair0537:27658:27984 [3] NCCL INFO CUDA Dev 3, IB Ports : mlx5_1/1(PIX) mlx5_3/1(SOC) mlx5_0/1(PHB) mlx5_2/1(SOC)
learnfair0536:46393:46747 [0] NCCL INFO CUDA Dev 0, IB Ports : mlx5_1/1(PHB) mlx5_3/1(SOC) mlx5_0/1(PIX) mlx5_2/1(SOC)
learnfair0536:46396:46761 [3] NCCL INFO CUDA Dev 3, IB Ports : mlx5_1/1(SOC) mlx5_3/1(PIX) mlx5_0/1(SOC) mlx5_2/1(PHB)
learnfair0536:46393:46747 [0] NCCL INFO Using 256 threads
learnfair0536:46393:46747 [0] NCCL INFO Min Comp Cap 7

learnfair0537:27655:27985 [0] init.cu:347 NCCL WARN Error : ring 0 does not loop back to start (7 != 1)
learnfair0537:27655:27985 [0] NCCL INFO init.cu:505 -> 3
learnfair0537:27655:27985 [0] NCCL INFO init.cu:593 -> 3
learnfair0536:46393:46747 [0] NCCL INFO Ring 00 :    0   2   4   6   5   1   3   7

learnfair0536:46393:46747 [0] init.cu:347 NCCL WARN Error : ring 0 does not loop back to start (4 != 0)
learnfair0536:46393:46747 [0] NCCL INFO init.cu:505 -> 3
learnfair0536:46393:46747 [0] NCCL INFO init.cu:593 -> 3
learnfair0536:46393:46747 [0] NCCL INFO misc/group.cu:69 -> 3 [Async thread]

learnfair0536:46394:46762 [1] init.cu:347 NCCL WARN Error : ring 0 does not loop back to start (6 != 2)
learnfair0536:46394:46762 [1] NCCL INFO init.cu:505 -> 3
learnfair0536:46394:46762 [1] NCCL INFO init.cu:593 -> 3
learnfair0536:46394:46762 [1] NCCL INFO misc/group.cu:69 -> 3 [Async thread]

learnfair0536:46395:46758 [2] init.cu:347 NCCL WARN Error : ring 0 does not loop back to start (5 != 4)
learnfair0536:46395:46758 [2] NCCL INFO init.cu:505 -> 3
learnfair0536:46395:46758 [2] NCCL INFO init.cu:593 -> 3

learnfair0536:46396:46761 [3] init.cu:347 NCCL WARN Error : ring 0 does not loop back to start (1 != 6)
learnfair0536:46396:46761 [3] NCCL INFO init.cu:505 -> 3
learnfair0536:46396:46761 [3] NCCL INFO init.cu:593 -> 3

learnfair0537:27656:27988 [1] init.cu:347 NCCL WARN Error : ring 0 does not loop back to start (4 != 3)
learnfair0536:46395:46758 [2] NCCL INFO misc/group.cu:69 -> 3 [Async thread]

learnfair0537:27657:27989 [2] init.cu:347 NCCL WARN Error : ring 0 does not loop back to start (3 != 5)
learnfair0536:46396:46761 [3] NCCL INFO misc/group.cu:69 -> 3 [Async thread]

learnfair0537:27655:27985 [0] NCCL INFO misc/group.cu:69 -> 3 [Async thread]
learnfair0537:27656:27988 [1] NCCL INFO init.cu:505 -> 3
learnfair0537:27656:27988 [1] NCCL INFO init.cu:593 -> 3
learnfair0537:27657:27989 [2] NCCL INFO init.cu:505 -> 3
learnfair0537:27657:27989 [2] NCCL INFO init.cu:593 -> 3
learnfair0537:27657:27989 [2] NCCL INFO misc/group.cu:69 -> 3 [Async thread]
learnfair0537:27658:27984 [3] init.cu:347 NCCL WARN Error : ring 0 does not loop back to start (6 != 7)
learnfair0537:27658:27984 [3] NCCL INFO init.cu:505 -> 3
learnfair0537:27658:27984 [3] NCCL INFO init.cu:593 -> 3
learnfair0537:27656:27988 [1] NCCL INFO misc/group.cu:69 -> 3 [Async thread]
learnfair0537:27658:27984 [3] NCCL INFO misc/group.cu:69 -> 3 [Async thread]
```

**Comment by sjeaugey** [comment]
It seems on learnfair0537, GPUs 0-1 are close to NIC 0, while GPUs 2-3 are close to NIC1. This is expected and usual.

However, on learnfair0536, GPU 0 is close to NIC 0, GPU 1 is close to GPU 2 and GPUs 2-3 are close to NIC 3. It makes things harder for NCCL to create rings between nodes since the GPUs-NIC mappings are very different between nodes (it looks as if CUDA_VISIBLE_DEVICES=0,4,6,7 was set on that node). It also means that GPUs on learnfair0536 would not have good intra-node NVLink connectivity.

So maybe NCCL could build some ring out of that, but performance would likely be low. Can you confirm it is intended ?

**Comment by teng-li** [comment]
@sjeaugey This scenario is inevitable if the user would like to schedule 4 instead of 8 GPUs since it will get scheduled much faster.  So it is pretty much depending on the GPU availability within different nodes and thus can cause the randomness of CUDA_VISIBLE_DEVICES.

**Comment by sjeaugey** [comment]
OK thanks for the rationale. I don't know exactly why it fails (would need to dig into the code) but I'm not surprised it does given how complex it is to get the rings right in all corner cases.

Given that most random sets of 4 GPUS out of 8 would run at 5GB/s instead of 40-60GB/s, and would impact other jobs running on the same node (they need to go through shared memory to communicate), resource allocation systems usually allow sub-partitioning of nodes only on aligned set of GPUs (so when allocating 4 GPUs it would either get 0,1,2 and 3 or 4,5,6 and 7). But of course there is a trade-off between performance and resource usage.

Anyway, I'll see if there is an easy fix here, hopefully it doesn't mean rewriting the whole ring creation system :-)

**Comment by sjeaugey** [comment]
Fixed in 2.5.6.
